README.Rmd

```diff
@@ -23,82 +28,106 @@ repositories of validated packages with risk metrics.
 
 ## Installation
 
-``` r
+_using `devtools`_
+
+```r
 # install.packages("devtools")
 devtools::install_github("pharmaR/pharmapkgs")
 ```
 
-## Bare bones demo
+_using `pak`_
+
+```r
+# install.packages("pak")
+pak::pkg_install("github::pharmaR/pharmapkgs")
+```
+
+## Bare Bones Demo
 
 All available packages from local repository.
 
 ```{r}
 library(pharmapkgs)
 
-# A local repository, bundled with pharmapkgs.
-(repo <- bundled_repos("ubuntu-22.04"))
+# local snapshot, bundled with pharmapkgs
+repo <- bundled_repos("ubuntu-22.04")
 
-# No filtering.
-all_pkgs <-
-  available.packages(contriburl = repo,
-                     fields = risk_fields(repo)) |>
+available.packages(repos = repo, fields = risk_fields(repo)) |>
   tibble::as_tibble()
-
-dplyr::glimpse(all_pkgs)
 ```
 
 Packages filtered using `{riskmetrics}` criteria.
 
 ```{r}
-# Fancy criteria. 
-my_filters <- risk_filter(ReverseDependencies < 0.5, HasNews == "1", RemoteChecks > 0.9)
-
-# Validated packages.
-val_pkgs <-
-  available.packages(contriburl = repo,
-                     fields = risk_fields(repo),
-                     filters = my_filters) |>
-  tibble::as_tibble()
+# Fancy criteria.
+my_filters <- risk_filter(
+  ReverseDependencies < 0.5,
+  HasNews == "1",
+  RemoteChecks > 0.9
+)
 
-dplyr::glimpse(val_pkgs)
+# Apply filter when inspecting packages
+tibble::as_tibble(available.packages(
+  repos = repo,
+  fields = risk_fields(repo),
+  filters = my_filters
+))
 ```
 
-Currently, not working, `pak::pkg_install()` is not downloading packages
-based on the DownloadURL field from the local repository.
+## Filtered Available Packages Workflow
+
+Configure a session to leverage a bundled repo.
 
 ```{r}
-if ("munsell" %in% rownames(installed.packages()))
-  pak::pkg_remove("munsell")
+repo <- bundled_repos("ubuntu-22.04")
+options(
+  # importantly, repo is called CRAN such that pak doesn't insert a CRAN mirror
+  repos = c(CRAN = repo)
+)
 
-pak::cache_clean()
-options(repos = repo)
-getOption("repos")
+nrow(available.packages())
+```
 
-# Install one of the validated packages with pak.
-istats <- pak::pkg_install("standard::munsell") |>
-  tibble::as_tibble()
-  
-dplyr::glimpse(istats)
+For ease of use, modify `available.packages` to globally take a default
+value from `options(available_packages_fields)`, allowing the use of those
+fields within `risk_filter()`.
 
-istats$sources
+This allows `pak` to leverage the same internal mechanisms for fetching
+packages and ensuring, at the time of download, that packages adhere to the 
+filters that we specify.
 
-istats$ref
+```{r}
+formals(available.packages)$fields <-
+  quote(getOption("available_packages_fields", NULL))
```

Comment:
This is a bit of a hack, but is pretty benign as far as base function modification goes.

However, we can avoid modifying `base::available.packages()` entirely if we do some additional legwork in our `risk_filter()` function.

The underlying challenge is that generally `available.packages` will not include `fields = risk_fields`, but we could call `available.packages()` from _within_ our `risk_filter` function and provide the necessary fields. We just have to make sure we don't accidentally infinitely recurse. 

Since this would introduce a fair bit of new complexity, I opted for a temporary hacky solution just to prove out that we can get `pak` working as intended.

Reply:
Nice touch @dgkf!